src/v/cluster/tests/shard_placement_table_test.cc

```diff
@@ -0,0 +1,478 @@
+    model::shard_revision_id cur_shard_revision{1};
+    raft::group_id cur_group{1};
+
+    co_await start();
+
+    for (size_t i = 0; i < 10'000; ++i) {
+        if (random_generators::get_int(15) == 0) {
+            vlog(clusterlog.info, "waiting for reconciliation");
+            for (size_t i = 0;; ++i) {
+                ASSERT_TRUE_CORO(i < 50) << "taking too long to reconcile";
+                if (!co_await rb.local().is_reconciled()) {
+                    co_await ss::sleep(100ms);
+                } else {
+                    break;
+                }
+            }
+            vlog(clusterlog.info, "reconciled");
+        }
+
+        // small set of ntps to ensure frequent overlaps
+        model::ntp ntp(
+          model::kafka_namespace, "test_topic", random_generators::get_int(10));
+
+        std::optional<shard_placement_target> target;
+
+        auto pt_it = ntpt.local().ntp2meta.find(ntp);
+        if (pt_it == ntpt.local().ntp2meta.end()) {
+            // add
+            auto group = cur_group++;
+            auto revision = cur_revision++;
+            co_await ntpt.invoke_on_all([&](ntp_table& ntpt) {
+                ntpt.ntp2meta[ntp] = ntp_table::ntp_meta{
+                  .group = group,
+                  .log_revision = revision,
+                };
+                ntpt.revision = revision;
+            });
+
+            target = shard_placement_target(
+              revision, random_generators::get_int(get_max_shard_id()));
+        } else {
+            auto ntp_meta = pt_it->second;
+
+            enum class op_t {
+                transfer,
+                remove,
+                increase_log_rev,
+            };
+
+            op_t op = random_generators::random_choice(
+              {op_t::transfer, op_t::remove, op_t::increase_log_rev});
+            switch (op) {
+            case op_t::transfer:
+                target = shard_placement_target(
+                  ntp_meta.log_revision,
+                  random_generators::get_int(get_max_shard_id()));
+                break;
+            case op_t::remove: {
+                auto revision = cur_revision++;
+                target = std::nullopt;
+                co_await ntpt.invoke_on_all([&](ntp_table& ntpt) {
+                    ntpt.ntp2meta.erase(ntp);
+                    ntpt.revision = revision;
+                });
+                break;
+            }
+            case op_t::increase_log_rev:
+                ntp_meta.log_revision = cur_revision++;
+                target = shard_placement_target(
+                  ntp_meta.log_revision,
+                  random_generators::get_int(get_max_shard_id()));
+                co_await ntpt.invoke_on_all([&](ntp_table& ntpt) {
+                    ntpt.ntp2meta[ntp] = ntp_meta;
+                    ntpt.revision = ntp_meta.log_revision;
+                });
+                break;
+            }
+        }
+
+        co_await spt.local().set_target(
```

Comment:
i can't quite tell, but it looks like concurrent partition movements are allowed and occur in the stress test. is that true?

Reply:
In the test a new shard assignment or a new log revision can be introduced concurrently while an old transfer is still ongoing. For several log revisions transfers can be concurrent.